Infer metadata from the data *.csv file

Bachibouzouk opened the pull request:
# Description

The function infer_metadata_from_data can be run on a datapackage where only the data/elements and data/sequences have been defined by a user and will generate automatically the datapackage.json file containing the metadata.

## Type of change

Please tick or delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Checklist:

Please tick or delete options that are not relevant.

- [x] New and adjusted code is formatted using the `pre-commit` hooks
 --> I couldn't get the pre-commit `local` tagged object to run, it told me `isort` is unknown command even after I pip installed it ....
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] existing unit tests pass locally with my changes
- [x] I have added new features/fixes to the [CHANGELOG](https://github.com/oemof/oemof-tabular/tree/dev/CHANGELOG.rst)
- [x] I have added my name to [AUTHORS]((https://github.com/oemof/oemof-tabular/tree/dev/AUTHORS.rst)
)

Bachibouzouk committed:
Map fields to the keys of FOREIGN_KEY_DESCRIPTORS

Each key map to a list of descriptors which themselves contain a
"fields" attribute.
We map the values under this "fields" attribute to the keys of
FOREIGN_KEY_DESCRIPTORS

Bachibouzouk committed:
Add function to infer datapackage foreign_keys

The function infer_metadata_from_data can be run on a datapackage where
only the data/elements and data/sequences have been defined by a user
and will generate automatically the datapackage.json file containing the
metadata.

Bachibouzouk committed:
Update documentation

Bachibouzouk commented on src/oemof/tabular/datapackage/building.py:
```diff
@@ -59,6 +60,178 @@ def update_package_descriptor():
+            p.add_resource(r.descriptor)
+
+
+def infer_metadata_from_data(
```
Idea was to make this feature optional, therefore I did not want to modify `infer_metadata` function

Bachibouzouk commented on src/oemof/tabular/datapackage/building.py:
```diff
@@ -59,6 +60,178 @@ def update_package_descriptor():
+        if "/elements/" in r.descriptor["path"]:
+            infer_resource_basic_foreign_keys(r)
+    # this function saves the metadata of the package in json format
+    infer_metadata(
```
Because this function does already part of the job if provided a dict `foreign_keys`, I used it and therefore just wrote the function ` infer_resource_basic_foreign_keys` to fill this `foreign_keys` dict.

Bachibouzouk committed:
Update oemof.tabular version

Bachibouzouk committed:
Fix flake8

Bachibouzouk committed:
Update AUTHORS.rst

Bachibouzouk committed:
Update CHANGELOG.rst

henhuy commented on src/oemof/tabular/datapackage/building.py:
```diff
@@ -59,6 +60,187 @@ def update_package_descriptor():
+
+def infer_metadata_from_data(
+    package_name="default-name",
+    path=None,
```
Default parameter "None" throws an error and makes no sense IMO.
I would make it mandatory and first argument.

Bachibouzouk replied:
Fixed in [0c4279b](https://github.com/oemof/oemof-tabular/pull/173/commits/0c4279bc54abbc8fe113d3d0bb358bb7c178af58)[0c4279b](https://github.com/oemof/oemof-tabular/pull/173/commits/0c4279bc54abbc8fe113d3d0bb358bb7c178af58)

henhuy commented on src/oemof/tabular/datapackage/building.py:
```diff
@@ -59,6 +60,187 @@ def update_package_descriptor():
+    metadata_filename="datapackage.json",
+):
+    """
+
```
Small docstring would be good

Bachibouzouk replied:
Fixed in [0c4279b](https://github.com/oemof/oemof-tabular/pull/173/commits/0c4279bc54abbc8fe113d3d0bb358bb7c178af58)[0c4279b](https://github.com/oemof/oemof-tabular/pull/173/commits/0c4279bc54abbc8fe113d3d0bb358bb7c178af58)

henhuy commented on src/oemof/tabular/datapackage/building.py:
```diff
@@ -59,6 +60,187 @@ def update_package_descriptor():
+            sequence_labels += field_labels
+
+    if duplicated_labels:
+        # write an error message here
+        pass
```
An error should be thrown here.

Bachibouzouk replied:
Fixed in [b8b4e64](https://github.com/oemof/oemof-tabular/pull/173/commits/b8b4e64301ff21d3e13990db231d9ee3588160ba)[b8b4e64](https://github.com/oemof/oemof-tabular/pull/173/commits/b8b4e64301ff21d3e13990db231d9ee3588160ba)

henhuy reviewed (changes requested):
Nice feature! Thanks for implementing.
Tested it with existing investment datapackage from examples - files were almost identical (only no empty line add end of file was missing :stuck_out_tongue:).
Would be nice to such test in tests folder too (aka build metadata for existing datapackage in examples and check if files are identical).
I also found three smaller issues.

Bachibouzouk commented:
> Nice feature! Thanks for implementing. Tested it with existing investment datapackage from examples - files were almost identical (only no empty line add end of file was missing 😛). Would be nice to such test in tests folder too (aka build metadata for existing datapackage in examples and check if files are identical). I also found three smaller issues.

Thanks for your review and your comments :) 

I will implement your suggestions and run those on the example datapackages :)

Bachibouzouk commented on src/oemof/tabular/datapackage/building.py:
```diff
@@ -59,6 +60,187 @@ def update_package_descriptor():
+    sequence_labels = []
+    duplicated_labels = []
+    for r in p.resources:
+        if "/sequences/" in r.descriptor["path"]:
```
Here one has to use `os.sep` instead of `"/"`

Bachibouzouk commented on src/oemof/tabular/datapackage/building.py:
```diff
@@ -59,6 +60,187 @@ def update_package_descriptor():
+    sequences_profiles_to_resource = map_sequence_profiles_to_resource_name(p)
+
+    for r in p.resources:
+        if "/elements/" in r.descriptor["path"]:
```
Here one has to use `os.sep` instead of `"/"`

Bachibouzouk replied:
Fixed in [37c5ad5](https://github.com/oemof/oemof-tabular/pull/173/commits/37c5ad564fc49a5ae7dc7e573df0c93e7717256f)[37c5ad5](https://github.com/oemof/oemof-tabular/pull/173/commits/37c5ad564fc49a5ae7dc7e573df0c93e7717256f)

Bachibouzouk committed:
Make path argument mandatory

Bachibouzouk committed:
Fix path separator bug

Bachibouzouk committed:
Raise error message in case of non unique sequences labels

Bachibouzouk committed:
Lint with black

Bachibouzouk commented:
> Would be nice to such test in tests folder too (aka build metadata for existing datapackage in examples and check if files are identical).

I did run the new function on files and some of them are just changing the order of the resources within the datapackage. There was a difference which helped me find a bug and solve it in c7ab4bfe106bba2ab1acc557e34ae705111f77ac

Bachibouzouk committed:
Fix flake error

Bachibouzouk committed:
Fix foreign key parsing bug

Bachibouzouk committed:
Update example datapackages

Those were generated by infer_metadata_from_data
and had no differences except the package name

Bachibouzouk committed:
Update example datapackages

Those were generated by infer_metadata_from_data
and had only a different order of resources

Bachibouzouk committed:
Sort resources alphabetically

Bachibouzouk committed:
Fix failing test

henhuy reviewed (approved):
LGTM! Thx for the changes
